### ISSUE TITLE
Odd behaviour of contract

### ISSUE THREAD

**Issue by kieranelby** [kieranelby/KingOfTheEtherThrone#1]
See https://www.reddit.com/r/ethereum/comments/44h1m1/a_new_%C3%B0app_king_of_the_ether_throne/czqyomf

**Comment by kieranelby** [comment]
Have created exactly the same contract on the testnet, seems to be working fine:

```
var contractAddr = '0x5fb32729f9f724e56e27c18dd8dbb6a63f759457';
var kingoftheetherthroneContract = web3.eth.contract([{"constant":false,"inputs":[{"name":"amount","type":"uint256"}],"name":"sweepCommission","outputs":[],"type":"function"},{"constant":true,"inputs":[],"name":"currentClaimPrice","outputs":[{"name":"","type":"uint256"}],"type":"function"},{"constant":true,"inputs":[{"name":"","type":"uint256"}],"name":"pastMonarchs","outputs":[{"name":"etherAddress","type":"address"},{"name":"name","type":"string"},{"name":"claimPrice","type":"uint256"},{"name":"coronationTimestamp","type":"uint256"}],"type":"function"},{"constant":false,"inputs":[{"name":"name","type":"string"}],"name":"claimThrone","outputs":[],"type":"function"},{"constant":true,"inputs":[],"name":"numberOfMonarchs","outputs":[{"name":"n","type":"uint256"}],"type":"function"},{"constant":true,"inputs":[],"name":"currentMonarch","outputs":[{"name":"etherAddress","type":"address"},{"name":"name","type":"string"},{"name":"claimPrice","type":"uint256"},{"name":"coronationTimestamp","type":"uint256"}],"type":"function"},{"constant":false,"inputs":[{"name":"newOwner","type":"address"}],"name":"transferOwnership","outputs":[],"type":"function"},{"inputs":[],"type":"constructor"},{"anonymous":false,"inputs":[{"indexed":false,"name":"usurperEtherAddress","type":"address"},{"indexed":false,"name":"usurperName","type":"string"},{"indexed":false,"name":"newClaimPrice","type":"uint256"}],"name":"ThroneClaimed","type":"event"}]);
var kingoftheetherthrone = kingoftheetherthroneContract.at(contractAddr);

// starting balances

> eth.getBalance(eth.accounts[2]);
10000000000000000000
> eth.getBalance(eth.accounts[3]);
5000000000000000000

// claim the empty throne

> eth.sendTransaction({
......   from: eth.accounts[2],
......   value: web3.toWei(100, 'finney'), // change this
......   to: '0x5fb32729f9f724e56e27c18dd8dbb6a63f759457',
......   gas: 500000,
......   data: web3.fromAscii('TestNet1') // change this
...... });

> eth.getBalance(contractAddr);
100000000000000000

> eth.sendTransaction({
......   from: eth.accounts[3],
......   value: web3.toWei(150, 'finney'),
......   to: '0x5fb32729f9f724e56e27c18dd8dbb6a63f759457',
......   gas: 500000,
......   data: web3.fromAscii('TestNet2')
...... });

> eth.getBalance(eth.accounts[2]);
10040653950000000000
> eth.getBalance(eth.accounts[3]);
4842559950000000000
> eth.getBalance(contractAddr);
101500000000000000

eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(225, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet3')});

> eth.getBalance(eth.accounts[2]);
9808213900000000000
> eth.getBalance(eth.accounts[3]);
5065309950000000000
> eth.getBalance(contractAddr);
103750000000000000
>

> kingoftheetherthrone.currentClaimPrice();
337000000000000000

// deliberately send too little
eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(300, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet4')});

> eth.getBalance(eth.accounts[2]);
9806768850000000000
> eth.getBalance(eth.accounts[3]);
5065309950000000000
> eth.getBalance(contractAddr);
103750000000000000

eth.sendTransaction({from: eth.accounts[3], value: web3.toWei(337, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet5')});

> eth.getBalance(eth.accounts[2]);
10140398850000000000
> eth.getBalance(eth.accounts[3]);
4720869900000000000
> eth.getBalance(contractAddr);
107120000000000000

// deliberately send too much
eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(666, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet6')});

> kingoftheetherthrone.currentClaimPrice();
757000000000000000

> eth.getBalance(eth.accounts[2]);
9627613250000000000

> eth.getBalance(eth.accounts[3]);
5220819900000000000

> eth.getBalance(contractAddr);
112170000000000000

// addr2 went down by 505 finney + some gas
10140398850000000000 - 9627613250000000000;
=512785600000000000

// addr3 went down by 505 finney - the 1% commission
5220819900000000000 - 4720869900000000000;
=499950000000000000

// contract balance went up by 1% of 505 finney
112170000000000000 - 107120000000000000;
=5050000000000000

// do a normal one
eth.sendTransaction({from: eth.accounts[3], value: web3.toWei(757, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet7')});

> kingoftheetherthrone.currentClaimPrice();
1130000000000000000
/// argh forgot to get balance of account 2!
> eth.getBalance(eth.accounts[3]);
4456378550000000000
> eth.getBalance(contractAddr);
119740000000000000

// how about a zero transaction
eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(0, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet8')});

> kingoftheetherthrone.currentClaimPrice();
1130000000000000000
> eth.getBalance(eth.accounts[2]);
10375933200000000000
> eth.getBalance(eth.accounts[3]);
4456378550000000000
> eth.getBalance(contractAddr);
119740000000000000

// how about another zero transaction
eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(0, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet9')});

// seemed ok, just lost some gas
> eth.getBalance(eth.accounts[2]);
10374823150000000000
> eth.getBalance(eth.accounts[3]);
4456378550000000000
> eth.getBalance(contractAddr);
119740000000000000
> kingoftheetherthrone.currentClaimPrice();
1130000000000000000

// do a normal one
eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(1130, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet10')});

> eth.getBalance(contractAddr);
119740000000000000
> eth.getBalance(eth.accounts[2]);
9237378400000000000
> eth.getBalance(eth.accounts[3]);
5575078550000000000
> eth.getBalance(contractAddr);
131040000000000000
> kingoftheetherthrone.currentClaimPrice();
1690000000000000000

// two people (almost exactly) same time

eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(1690, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet11')});
eth.sendTransaction({from: eth.accounts[3], value: web3.toWei(1690, 'finney'), to: contractAddr, gas: 500000, data: web3.fromAscii('TestNet12')});

> eth.getBalance(eth.accounts[2]);
9213033650000000000
> eth.getBalance(eth.accounts[3]);
5573630100000000000
> eth.getBalance(contractAddr);
147940000000000000
> kingoftheetherthrone.currentClaimPrice();
2530000000000000000

// try using less gas

eth.sendTransaction({from: eth.accounts[3], value: web3.toWei(2530, 'finney'), to: contractAddr, gas: 150000, data: web3.fromAscii('TestNet13')});

> eth.getBalance(eth.accounts[2]);
11717733650000000000
> eth.getBalance(eth.accounts[3]);
3036185350000000000
> eth.getBalance(contractAddr);
173240000000000000
> kingoftheetherthrone.currentClaimPrice();
3790000000000000000

// try using even less gas

eth.sendTransaction({from: eth.accounts[2], value: web3.toWei(2530, 'finney'), to: contractAddr, gas: 30000, data: web3.fromAscii('TestNet14')});

> eth.getBalance(eth.accounts[2]);
11716233650000000000
> eth.getBalance(eth.accounts[3]);
3036185350000000000
> kingoftheetherthrone.currentClaimPrice();
3790000000000000000
> eth.getBalance(contractAddr);
173240000000000000
```

**Comment by kieranelby** [comment]
some things I haven't tried:

- using long strings or non-ascii strings
- commission sweeping
- claiming the throne from the "wizard" adddress

**Comment by kieranelby** [comment]
ok, that's even more weird, https://etherchain.org/account/0xb336a86e2feb1e87a328fcb7dd4d04de3df254d0 says the contract has a negative balance of -0.71533 Ether !!

**Comment by kieranelby** [comment]
Added some more info  to https://www.reddit.com/r/ethereum/comments/44h1m1/a_new_%C3%B0app_king_of_the_ether_throne/czvdjl4

**Comment by kieranelby** [comment]
Wrote up a post-mortem of the issue here: http://www.kingoftheether.com/postmortem.html

**Comment by kieranelby** [comment]
Committed initial fix, needs some more testing.

**Comment by kieranelby** [comment]
Also need to fix long string problem

**Comment by kieranelby** [comment]
Fixed in 1.0
